libsolidity/ast/Types.cpp

```diff
@@ -314,7 +314,7 @@ char const* Type::categoryName(Type::Category _category)
 	case Category::UserDefinedValueType: return "user-defined value type";
 	case Category::Tuple: return "tuple";
 	case Category::Mapping: return "mapping";
-	case Category::TypeType: return "type of a type";
+	case Category::TypeType: return "type";
```

Comment:
No, it doesn't, it's the type of the struct type... ``S`` is not a struct, it's the type of a struct...

Reply:
These are *type* categories, the categories of *types* if expressions, nit categories of expressions - you're thinking on the wrong layer here :-).
So this is definitely still the type of a type and **not** a type. Why would it be called ``TypeType`` otherwise :-)?

Reply:
> And you use S in an expression, its AST node is Category::TypeType. So it refers to a type (a struct type).

To try to clarify this: this is incorrect - the AST node is an expression :-). The expression *is* itself a type, but that's not the category of the type of the expression - the type of this expression is the type of a type (since the expression is a type). The translation of categories of types to categories of expression was abusive in #14087 already.

Reply:
But yeah, this is primarily a distraction on this PR - I'd suggest to just remove the change here and then we can argue about this separately. But I'd say we should just revert #14087 and solve this that way, everything else is too much time spent on this.

Reply:
> These are type categories, the categories of types of expressions, not categories of expressions themselves - you're thinking on the wrong layer here :-).

You're right. I got that wrong. I was thinking about this:

```solidity
1;          // RationalNumberType, Category::RationalNumber
x;          // IntegerType,        Category::Integer
uint;       // TypeType,           Category::TypeType
type(uint); // MagicType,          Category::Magic

s;          // StructType,         Category::Struct
S;          // TypeType,           Category::TypeType

c;          // ContractType,       Category::Contract
C;          // TypeType,           Category::TypeType
type(C);    // MagicType,          Category::Magic
```

but these indeed describe the type of these things and not the things themselves.

See also https://github.com/ethereum/solidity/pull/12656#discussion_r1170020290.

Reply:
> But I'd say we should just revert https://github.com/ethereum/solidity/pull/14087 and solve this that way, everything else is too much time spent on this.

I will do that (separately). I now see what you meant earlier. I'll drop the `categoryName()` function.

Reply:
Yeah, the cause for the confusion may be that for brevity ``Type::Category::*`` does not repeat ``Type`` at the end explicitly in all cases, since given the scope it's implicit - i.e. ``Type::Category::Integer`` is the category of *types* that are integer *types* - and only ``TypeType`` spells that out again since ``Category::Type`` would more likely give the wrong impression :-). So it's not exactly perfect, but now that this is cleared up, we don't have to do anything about it other than https://github.com/ethereum/solidity/pull/12656#discussion_r1170072290

We'll need to rethink all this a bit in our upcoming new type system anyways.